explorer/data/prelude.carbon

```diff
@@ -416,6 +416,48 @@ external impl i32 as RightShiftWith(i32) where .Result == i32 {
   }
 }
 
+//-------------------------
+// Optional
+//-------------------------
+choice OptionalElement(T:! Type) {
+  None(),
+  Element(T)
+}
+
+class Optional(T:! Type){
+  fn CreateEmpty() -> Optional(T){
+    return { .element = OptionalElement(T).None() };
+  }
+  fn Create(value: T ) -> Optional(T){
+    return { .element = OptionalElement(T).Element(value) };
+  }
+
+  fn HasValue[me: Self]() -> bool {
+    match(me.element){
+      case OptionalElement(T).None() => { return false; }
+    }
+    return true;
+  }
+
+  fn Get[me: Self]() -> T {
+    var y: T;
+    match(me.element){
+      case OptionalElement(T).Element(x: T ) =>{
+        return x;
+      }
+    }
+    return y;
+  }
+
+  impl Optional(T) as ImplicitAs(bool){
+    fn Convert[me: Self]() -> bool{
+      return me.HasValue();
+    }
+  }
```

Comment:
This doesn't follow our design principles for implicit conversions, because it isn't [lossless](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/implicit_conversions.md#lossless). Let's leave it out.

Reply:
Really, We want only allow to test Optional in this way?
```carbon

var opt: Optional(i32) =  Optional.Create(2);

if(opt.HasValue()){
...
}
```
it's ok for me, but it seems quite uncomfortable at first, doesn't it ?

Reply:
It doesn't seem uncomfortable to me, but I understand that it could be uncomfortable to people who are more accustomed to using implicit boolean conversions. I think it's the right tradeoff, though, because these conversions are dangerous. I've seen very serious production outages caused by code that said `if(x)` when the author meant `if(*x)`, and a couple years back an engineer at Google audited our uses of `absl::StatusOr<bool>::operator bool`, and found that roughly 75% of them were bugs of the same kind (for that reason, `absl::StatusOr` no longer has an `operator bool`).

In any event, whenever we have a design proposal for `Optional`, it should address this issue. In the meantime, we should stick to the documented principles for implicit conversions.

Reply:
All clear...! It's fine

Reply:
Done!